include/irr/asset/filters/CSummedAreaTableImageFilter.h

```diff
@@ -0,0 +1,256 @@
+				{
+					constexpr auto decodeByteSize = 8u;
+					auto channels = asset::getFormatChannelCount(format);
+					return channels * decodeByteSize * extent.width * extent.height * extent.depth;
+				}
+		};
+
+	protected:
+		CSummedAreaTableImageFilterBase() {}
+		virtual ~CSummedAreaTableImageFilterBase() {}
+
+		static inline bool validate(CSummStateBase* state)
+		{
+			if (!state)
+				return false;
+
+			return true;
+		}
+};
+
+//! Fill texel buffer with computed sum of left and down texels placed in input image
+/*
+	When the summing is in exclusive mode - it computes the sum of all the pixels placed
+	on the left and down for a new single texel but it doesn't take sum the main texel itself.
+	In inclusive mode, the texel we start from is taken as well and added to the sum.
+*/
+
+template<bool ExclusiveMode = false>
+class CSummedAreaTableImageFilter : public CMatchedSizeInOutImageFilterCommon, public CSummedAreaTableImageFilterBase<ExclusiveMode>
+{
+	public:
+		virtual ~CSummedAreaTableImageFilter() {}
+
+		class CStateBase : public CMatchedSizeInOutImageFilterCommon::state_type, public CSummedAreaTableImageFilterBase<ExclusiveMode>::CSummStateBase {};
+		using state_type = CStateBase; //!< full combined state
+
+		static inline bool validate(state_type* state)
+		{
+			if (!CMatchedSizeInOutImageFilterCommon::validate(state))
+				return false;
+
+			if (!CSummedAreaTableImageFilterBase<ExclusiveMode>::validate(state))
+				return false;
+			
+			const ICPUImage::SCreationParams& inParams = state->inImage->getCreationParameters();
+			const ICPUImage::SCreationParams& outParams = state->outImage->getCreationParameters();
+			const auto inFormat = inParams.format;
+			const auto outFormat = outParams.format;
+
+			if (state->scratchMemoryByteSize < state_type::getRequiredScratchByteSize(outFormat, {outParams.extent.width, outParams.extent.height, outParams.extent.depth}))
+				return false;
+
+			if (asset::getFormatChannelCount(outFormat) < asset::getFormatChannelCount(inFormat))
+				return false;
+
+			if (asset::getFormatClass(state->inImage->getCreationParameters().format) < asset::getFormatClass(outFormat))
+				return false;
+
+			return true;
+		}
+
+		static inline bool execute(state_type* state)
+		{
+			if (!validate(state))
+				return false;
+
+			auto checkFormat = state->inImage->getCreationParameters().format;
+			if (isIntegerFormat(checkFormat) || isSRGBFormat(checkFormat))
+				return executeInterprated(state, reinterpret_cast<uint64_t*>(state->scratchMemory));
+			else
+				return executeInterprated(state, reinterpret_cast<double*>(state->scratchMemory));
+		}
+
+	private:
+
+		template<typename decodeTypePointer> //!< double or uint64_t
+		static inline bool executeInterprated(state_type* state, decodeTypePointer scratchMemory)
+		{
+			using decodeType = typename std::decay<decltype(*scratchMemory)>::type;
```

Comment:
u could just `remove_pointer_t<decodeTypePointer>` or even 
```
template<typename decodeType> //!< double or uint64_t
static inline bool executeInterprated(state_type* state, decodeType* scratchMemory)
```

Reply:
you're right, corrected!<3